Pidgin/Parser.Not.cs

```diff
@@ -0,0 +1,58 @@
+using System.Collections.Generic;
+using System.Linq;
+using Pidgin.ParseStates;
+
+namespace Pidgin
+{
+    public static partial class Parser
+    {
+        /// <summary>
+        /// Creates a parser which attempts to apply the specified parser and only succeeds if that parser fails.
+        /// </summary>
+        /// <param name="parser">The parser to apply if the current parser fails without consuming any input</param>
+        /// <returns>A parser which tries to apply the current parser, and applies <paramref name="parser"/> if the current parser fails without consuming any input.</returns>
+        public static Parser<TToken, TToken> Not<TToken>(Parser<TToken, TToken> parser) {
+            return new NegatedParser<TToken>(parser);
+        }
+        
+        private sealed class NegatedParser<TToken> : Parser<TToken, TToken>
+        {
+            private readonly Parser<TToken, TToken> _parser;
+
+            public NegatedParser(Parser<TToken, TToken> parser) : base(parser.Expected)
+            {
+                _parser = parser;
+            }
+
+            internal sealed override Result<TToken, TToken> Parse(IParseState<TToken> state)
+            {
+                var startingPosition = state.SourcePos;
+                var result = _parser.Parse(state);
+                if (result.Success)
+                {
+                    return Result.Failure<TToken, TToken>(
+                            new ParseError<TToken>(
+                                new Maybe<TToken>(result.Value),
+                                false,
+                                null,
+                                startingPosition,
+                                "unexpected token"
+                            ),
+                            true
+                        );
+                }
+
+                var token = state.Peek();
+                if (token.HasValue) {
+                    state.Advance();
+                    return Result.Success<TToken, TToken>(
+                            token.Value,
+                            true
+                        );
+                }
```

Comment:
> Can you think of any real world use cases where you would want it to do what your suggested test is doing?

It's the "principle of least surprise". I'd rather not make the decision to consume a character on behalf of the user, because I don't know what they're going to use `Not` for. If they _do_ want to start consuming characters after the wrapped parser fails, they can use `Then`; if they want to backtrack they can use `Try`.